src/main/java/com/redhat/devtools/lsp4ij/features/navigation/LSPGotoDeclarationHandler.java

```diff
@@ -42,25 +47,53 @@ public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
 
     @Nullable
     @Override
-    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {
+    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement,
+                                                  int offset,
+                                                  Editor editor) {
         Project project = editor.getProject();
         if (project == null || project.isDisposed()) {
             return PsiElement.EMPTY_ARRAY;
         }
-        PsiFile psiFile = sourceElement.getContainingFile();
+        PsiFile psiFile = sourceElement != null ? sourceElement.getContainingFile() : null;
         if (psiFile == null) {
             return PsiElement.EMPTY_ARRAY;
         }
         if (!LanguageServersRegistry.getInstance().isFileSupported(psiFile)) {
             return PsiElement.EMPTY_ARRAY;
         }
+
+        // If this was called for a populated semantic tokens view provider, just get the target directly from it
+        if (psiFile.getViewProvider() instanceof LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider) {
+            if (semanticTokensFileViewProvider.isReference(offset)) {
+                PsiReference reference = semanticTokensFileViewProvider.findReferenceAt(offset);
+                PsiElement target = reference != null ? reference.resolve() : null;
+                return target != null ? new PsiElement[]{target} : PsiElement.EMPTY_ARRAY;
+            } else if (semanticTokensFileViewProvider.isDeclaration(offset)) {
+                return PsiElement.EMPTY_ARRAY;
+            }
+
+            // NOTE: It's important that we short-circuit any further processing of this potential reference if this
+            // file is backed by semantic tokens and it's not a reference or a declaration. Otherwise things that can't
+            // act as references will show up as navigatable/hyperlinked incorrectly.
+            throw new ProcessCanceledException();
+        }
+
+        // Otherwise use LSP directly
+        return getGotoDeclarationTargets(editor, sourceElement, offset);
+    }
+
+    @ApiStatus.Internal
+    public static PsiElement @NotNull [] getGotoDeclarationTargets(@NotNull Editor editor,
```

Comment:
Extracted the part that uses LSP to resolve the reference into a reusable method as we'll need it elsewhere as well.